### ISSUE TITLE
Generalize slicing and slice assignment ops (including gather and scatter)

### ISSUE THREAD

**Issue by girving** [tensorflow/tensorflow#206]
We should make our slicing and assignment ops more general to capture more of the functionality of numpy slicing, and add `__getitem__` sugar for all of it.  Specifically,

1. We should have a 2.5 dimensional set of ops, with dimensions (1) get vs. set, (2) slice type, and for the assignment ops (3) the update op.  Currently we have `slice`, `assign_update`, `assign_add`, `assign_sub`, `gather`, `scatter_update`, `scatter_add`, `scatter_sub`.  We should also have `assign_slice_update`, `assign_slice_add`, `assign_slice_sub`.

2. Both slicing and slice assignment should support strides, with no performance cost if strides aren't used.

3. Ideally, the slice ops should support negative indexing a la Python.  Since the slice parameters are already CPU, this is implementable with near zero cost.   The unfortunate bit is that since we picked the wrong format for specifying ranges (start + length instead of start : end), negative indexing might be awkward.  Thus, it might be best left to a separate bug.

4. Support numpy-like boolean indexing.

5. Generalize `gather` and `scatter_*` to take an array of input index tensors, efficiently broadcast them, and do multidimensional indexing similar to numpy.

6. Make `__getitem__` provide sugar for all of the above.  Ideally we'd have something idiomatically similar at least to `__setitem__`, but this is problematic since the returned assignment op is important to have, `__setitem__` does not return a value, and the nice range sugar is available only inside indexing / assignment calls.

@ebrevdo: I'm assigning this to you for now since you might get to it first, but feel free to grab only the piece of it that you need for now.

**Comment by girving** [comment]
Lasse requests the equivalent of numpy mixed indexing:

    x[:, tensor]

which combines slicing with indexing-by-tensor.

**Comment by lespeholt** [comment]
...where tensor can be either a scalar (which would select all the values in that column) or a vector which can select individual columns in the rows, so:
foo = tf.constant([[1,2,3], [4,5,6]])
foo[:, 1] # [2, 5]
indexes = tf.constant([1, 2])
foo[:, indexes] # [2, 6]

**Comment by avostryakov** [comment]
If I understand correctly the following code is exactly what is needed for cross-entropy loss function
indexes = tf.constant([1, 2])
foo[:, indexes] # [2, 6]

If we will have this kind of indexing we can write:
cost = -tf.reduce_sum(tf.log(network_output[:, targets]))

, where targets is a vector of class's indexes instead of

cost = -tf.reduce_sum(targets*tf.log(network_output))

, where targets is a sparse matrix

Am I correct?

**Comment by yaroslavvb** [comment]
Numpy also has newaxis and "Ellipsis" objects. IE, to prepend an axis using numpy notation
a[np.newaxis,...]

http://docs.scipy.org/doc/numpy-1.10.0/reference/arrays.indexing.html

**Comment by girving** [comment]
Yes, newaxis is essential.

**Comment by girving** [comment]
As part of this, we should make https://github.com/tensorflow/tensorflow/issues/418 work.

**Comment by olange-google** [comment]
It would also be helpful if gather supports an axis parameter: 

current behavior:
gather(v, indices)  --> output[i, ... ] = params[indices[i], ... ] 

wanted behavior: 
gather(v, indices, axis=1) --> output[:, i, :] = params[:, indices[i], :] 

(Please excuse, if this is covered by the list of requirements posted above already)

**Comment by girving** [comment]
@olange-google: I think we're unlikely to implement an axis parameter since it's beyond numpy features, but what you want is covered by the *combination* of slice indexing and advanced indexing.

**Comment by ebrevdo** [comment]
Are you sure?  If in numpy I use:

array[:, :, [1, 2, 3], :]

then that's equivalent to gather([1, 2, 3], axis=2)

is it not?

**Comment by girving** [comment]
We're saying the same thing.  I'm not objecting to that functionality, just to exposing it via that sort of axis parameter rather than as a special case of the combination of slice indexing and advanced indexing.

**Comment by mhejrati** [comment]
@ebrevdo What is the status on this issue? 
I am interested to implement this if no one is working on it.

**Comment by ebrevdo** [comment]
I plan to work on this but not until after next week. If you want to work
in this, I suggest supporting GPU and CPU in your kernels. If you can't
implement that, you may want to wait for us to implement it.

**Comment by MInner** [comment]
Hi :) Could you please give an update on the status of this feature?

The `_SliceHelper` docstring says that `the "stop" of the slice must not be omitted`, however this case seem to be handled just fine in function implementation:
```
if s.stop is None or s.stop == sys.maxsize:
        sizes.append(-1)
```
- or I'm getting something wrong?

**Comment by ebrevdo** [comment]
OK recently added the gather_nd op, which performs a special subset of the required functionality: given a tensor of indices, gather the requested values.

Advanced slicing is on the radar.

**Comment by waleedka** [comment]
@ebrevdo I tried using the gather_nd op to get the last relevant output from a variable length LSTM network. I'm passing sequence_length to the RNN, which means that the last few outputs of most examples are zeros, so I'm trying to read the last non-zero output. I'm getting this error, though, in the training phase:

**NotImplementedError: Gradient for gather_nd is not implemented.**

```
  outputs, state = rnn.rnn(multi_rnn_cell, inputs, dtype=tf.float32, sequence_length=lengths_ph)

  indicies = tf.concat(1, [
      tf.expand_dims(lengths_ph - 1, 1),
      tf.expand_dims(tf.range(tf.shape(vectors_ph)[0]), 1),
      tf.expand_dims(tf.zeros_like(lengths_ph), 1),
      ])
  output_tensor = tf.pack(outputs)
  relevant_output = tf.gather_nd(output_tensor, indicies)
```

**Comment by ebrevdo** [comment]
Yeah - we haven't written the gradient implementation for gather_nd yet.
It's essentially a reshape followed by a call to sparse_to_dense; but
sparse_to_dense doesn't have a GPU implementation (on my TODO) so I'm not
using it yet.

**Comment by hycis** [comment]
hi @ebrevdo, do you have any timeline to implement gradient for that, as we are currently using that function. and it's quite useful function.

**Comment by nova77** [comment]
While we wait for gather_nd for supporting gradients, this is a temporary solution:

```python
x = tf.constant([[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]])
idx = tf.constant([1, 0, 2])
idx_flattened = tf.range(0, x.shape[1]) * x.shape[0] + idx
y = tf.gather(tf.reshape(x, [-1]),  # flatten input
              idx_flattened)  # use flattened indices

with tf.Session(''):
  print y.eval()  # [2 4 9]
```

**Comment by ebrevdo** [comment]
I will implement a gradient in the next week.  Keep in mind that it will be
CPU-only for now.

On Wed, Apr 20, 2016 at 2:27 AM, Norman Casagrande <notifications@github.com

**Comment by danijar** [comment]
@nova77 Thanks for the example. What if `x` in your example has an additional dimension? I tried to adapt your example by reshaping to `tf.concat(shape[0] * shape[1], shape[2])` but apparently this is not a valid shape.

**Comment by erickrf** [comment]
@danijar this is a working solution, but when I tried it, I got the following warning from tensorflow:

UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "

Apparently, this is caused by the index slices returned by `tf.gather`. Does anyone know if there's a way to avoid this problem?

**Comment by danijar** [comment]
@erickrf Would be glad to hear of a better solution as well. Of course you can hide the warning as any other Python warning though.

**Comment by LeavesBreathe** [comment]
> I will implement a gradient in the next week.  Keep in mind that it will be
CPU-only for now.

Is this implemented yet by any chance @ebrevdo? 

From what I've seen, `tf.gather` is gpu gradient capable at least correct? Thanks @nova77  for the temporary solution!

**Comment by hycis** [comment]
# here is another implementation for 2d
def gather_2d(params, indices):
    # only for two dim now
    shape = params.get_shape().as_list()
    assert len(shape) == 2, 'only support 2d matrix'
    flat = tf.reshape(params, [np.prod(shape)])
    flat_idx = tf.slice(indices, [0,0], [shape[0],1]) * shape[1] + tf.slice(indices, [0,1], [shape[0],1])
    flat_idx = tf.reshape(flat_idx, [flat_idx.get_shape().as_list()[0]])
    return tf.gather(flat, flat_idx)

**Comment by keithshep** [comment]
@ebrevdo I could try to help out with `__getitem__` / `__setitem__` if it would be useful. I think these ops are going to be the most intuitive way for people to interact with tensors especially since most are already comfortable with how numpy works. Would you like me to try or would it be better to just leave you to implement it?

**Comment by ebrevdo** [comment]
@keithshep PRs are most welcome.  I would start with adding negative indexing to the current getitem/setitem (the one that currently works on the outermost dimension).  This will require changes to both the kernel and the python wrappers, and updates to the shape inference code.

**Comment by girving** [comment]
Cc @aselle.

**Comment by girving** [comment]
@keithshep: Do you mind if we give this to @aselle as a starter project?  He just joined Brain. :)

**Comment by keithshep** [comment]
@girving Not at all. I was only going to get a chance to dig in this weekend so no problem. Oh, and congrats to @aselle

**Comment by girving** [comment]
For `__setitem__`: we can't use it directly because it doesn't return a value, and usually you need the return value of assignments (to make sure they execute).  However, we could make `x[i].assign(y)` work with a bit of class trickery.

**Comment by keithshep** [comment]
@girving I don't know if this is a reasonable path or not (I may be missing the point) but I wonder if borrowing ideas from SSA https://en.wikipedia.org/wiki/Static_single_assignment_form for manipulating the computation graph in cases where there is a side effect and no return value is a way to achieve this?

**Comment by girving** [comment]
@keithshep: Unfortunately I don't know what you mean.  We can't manipulate a computation graph until we have one to manipulate, and `__setitem__` discards the output of the assign.  Normally that return value needs to be used later on, typically as part of a training op which is then run.

**Comment by keithshep** [comment]
@girving my thought was to add some tensor version metadata similar to variable versioning used in SSA. This would mean that some internal tensor version bookkeeping would have to be performed whenever `__setitem__` is invoked, and that a tensor version would have to be recorded whenever a tensor is passed to an op since a single tensor python reference could refer to different versions depending on when it's used. Not sure if that makes sense, but as I think about it more it's probably more complicated than is worthwhile anyway.

**Comment by girving** [comment]
@keithshep: That approach would be reasonable, but unfortunately I think it's too big a jump from what we have currently.

**Comment by shoyer** [comment]
Just a note -- we *do not* want to blindly copy NumPy vectorized indexing in `__getitem__`. Instead, we want so-called "outer indexing" (like MATLAB), which is much more intuitive.

This means that indexing with mixed slices and arrays *should* differ in TensorFlow compared to NumPy.

See this NumPy PR for examples and a discussion of the relevant issues: https://github.com/numpy/numpy/pull/6256/files

I'm happy to elaborate if that doesn't make sense.

**Comment by girving** [comment]
@shoyer: I don't think you need to elaborate.  Outer indexing makes sense, but it isn't what numpy does.  Numpy is also more flexible than outer indexing.  Finally, I like numpy indexing. :)  Thus, unless I see a bunch of support from other people for deviating from numpy, we'll stick with numpy.

**Comment by yaroslavvb** [comment]
@shoyer What do the odds look like for that PR being accepted on the Numpy side?

**Comment by shoyer** [comment]
@girving I agree that NumPy's vectorized indexing is great. The problem is that NumPy's implementation includes a hacks to make it "more intuitive" when slices and arrays are mixed, which work a lot of the time but then result in some very bizarre edge cases.

The [canonical example](https://mail.scipy.org/pipermail/numpy-discussion/2015-April/072607.html) is indexing operations like `x[0, :, [0, 1, 2, 3]]`, which converts an array of shape `(2, 3, 4)` to `(4, 3)` rather than the "obvious" `(3, 4)`. This is point (2) under the motivation in proposal PR (https://github.com/numpy/numpy/pull/6256).

One way to avoid this issue entirely is to prohibit mixed slice/array indexing. But if you do allow mixed slice/array indexing (which is very tempting to handle cases like `x[:, [0, 1, 2]]`), then this edge case will inevitably arise. Hence the appeal of the alternative indexing operation `vindex` for vectorized indexing, which lets indexing like `x[:, [0, 1, 2]]` and `x[[0, 1, 2], :]` do the obvious "outer indexing" thing without any kludges.

@yaroslavvb There are a few niggling details to figure out in the proposal, but I think it's extremely like that it (and the implementation https://github.com/numpy/numpy/pull/6075) will be merged at some point. There was a long discussion last year about just how unintuitive numpy's current indexing behavior is, and I think all the core developers agreed that this is the obvious path forward.

**Comment by girving** [comment]
@shoyer: Please get it merged on the numpy side and then file a separate issue.  I predict a 0% chance that numpy will remove the existing indexing support or make it not the default: that would break millions of lines of code and alienate huge numbers of numpy users.  If they add it as an optional feature, TensorFlow could easily do the same thing with the same syntax.

**Comment by danijar** [comment]
Will the version @aselle is working on support slicing by start/end tensors? This would be very useful for handling batches of variable-length sequences.

```python
data = [[2, 2, 1, 3], [3, 1, 0, 0], [2, 1, 3, 0]]
length = [4, 2, 3]
data[:, 1: length - 1] == [[2, 1], [], [1]]
```

**Comment by shampool** [comment]
When will the negative indices be supported? like a[:,-1]

**Comment by girving** [comment]
@danijar, @shampool: Yes on both counts.

**Comment by danijar** [comment]
@girving Okay, cool. I hate to spam the thread but is there a way to get the behavior from my example without the new syntax?

**Comment by girving** [comment]
@danijar: Let's keep this thread focused on the new feature.  Please ask side questions on StackOverflow.

**Comment by MInner** [comment]
hi, I attempted to write a temporary workaround (that someone might possibly find useful)
https://gist.github.com/MInner/8b0c0a0e528303b132bf02e277199996

it's more or less clear (at least in simplest case) how to do
`A[:, b, :] # => (10, 50, 30)` - transpose, gather, transpose back 
or 
`A[a, b, c] # => (50, )` - pack, gather_nd
, but it's not clear to me how could one do `A[a, b, :] # => (50, 30)` without creating super-large index matrix for `gather_nd`, though it's clear how to do `(50, 50, 30)`

`A is float [10, 20, 30] in (0..1); 
a, b, c are ints [50,] in [0..10]`

**Comment by ebrevdo** [comment]
I'm working on a gather_nd extension to support inner slicing.. Note that
gather_nd still has no gradient.

**Comment by gpapan** [comment]
A workaround for the special case of indexing across some dimension with a list of integers, which might be useful to some folks:

     ind = [3, 5, 0]
     # y = x[:,ind,:]  # this doesn't work right now
     y = tf.concat(1, [tf.expand_dims(x[:, i, :], 1) for i in ind])

I emphasize that this only works if `ind` a list of integerts but does not cover the case when `ind` is a Tensor of integers.

**Comment by cruvadom** [comment]
Inspired by @nova77, a workaround (until gather_nd has gradients), allowing to use gather_nd normally:

def gather_nd(params, indices, name=None):
  shape = params.get_shape().as_list()
  rank = len(shape)
  flat_params = tf.reshape(params, [-1])
  multipliers = [reduce(lambda x, y: x*y, shape[i+1:], 1) for i in range(0, rank)]
  indices_unpacked = tf.unpack(tf.transpose(indices, [rank - 1] + range(0, rank - 1), name))
  flat_indices = sum([a*b for a,b in zip(multipliers, indices_unpacked)])
  return tf.gather(flat_params, flat_indices, name=name)

**Comment by aselle** [comment]
- [x] Implement rvalue basic indexing
- [ ] Implement lvalue basic indexing
- [ ] Implement advanced indexing
- [ ] Implement bool mask indexing

**Comment by LiamConnell** [comment]
Here's another workaround for the case where your `params` variable is of unknown length. Hope it helps, and if there is a better way currently known please let me know! Not sure about differentiability because I'm using it for a policy gradient
``` 
# we want output[i] = params[i,indices[i]]
index_mask = tf.reshape(tf.one_hot(indices, 3), [-1,3])
output = tf.reduce_sum(params * index_mask,1)
```

**Comment by Fenugreek** [comment]
I am looking to do a max_pool that doesn't reduce the input tensor but replaces the non-max elements with zeros. Currently using a clumsy and slow workaround using space_to_depth, argmax, onehot, select and several transposes/reshapes.

Implementing numpy-like indexing for assignment will help simplify this, so I'd like to add my vote for this feature. If anyone has any other ideas on how I can go about this, let me know, Thanks.

**Comment by girving** [comment]
@Fenugreek: Please keep this thread focused on the new feature. Please ask side questions on StackOverflow.  Note that votes do not count as discussions of the feature.

**Comment by tejaskhot** [comment]
Will this also support multi-indexing such as:
`matrix[row_indices, col_indices]` where `matrix` is a `mxn` tensor and `row_indices` and `col_indices` are int32 vectors of sizes `k` such that k is less than m and n respectively. The slice above returns a length `k` vector eventually i.e. vectorized operation of `matrix[i,j] for all i,j in row_indices, col_indices`

**Comment by aselle** [comment]
Eventually we may add support for  that (it's what numpy calls advanced indexing), but this first version will just focus on making the so called numpy basic slicing fully supported.

**Comment by tejaskhot** [comment]
@aselle With what we have now, what's the optimal way of achieving this sort of indexing? I was hoping this would be solved in this issue.

**Comment by girving** [comment]
@tejaskhot: Please keep this thread focused on the new features. Please ask side questions on StackOverflow.

**Comment by cdjkim** [comment]
Any ideas as to when gather_nd will support gradient?

**Comment by hokyungLee** [comment]
we have 3 rank tensor([9, 7, 11)], so  we changed into 2 rank tensor([9,11]) with tf.gather_nd(...)
But the problem is, I've noticed that tensorflow optimizer can't support tf.gater_nd(..)
we were already applied to our program in above some idea
That's why, we want to get any idea
Please, help us....

**Comment by aselle** [comment]
Closing this issue. As of 0.10 improved basic indexing and sliced assignment is available. Advanced, mixed advanced and basic, as well as boolean slicing still need implementation. See #4638 and #4639.

**Comment by sonalgupta** [comment]
@aselle This issue was closed but the question of when will the gradient for gather_nd be added is still not answered.

**Comment by cruvadom** [comment]
@sonalgupta You can use my workaround from a previous comment.

**Comment by danijar** [comment]
@sonalgupta If I understand correctly, you don't need `tf.gather_nd()` anymore but can use slicing which has implemented gradients.

**Comment by aselle** [comment]
Internally  we are working on tf.scatter which will be used to implement tf.gather_nd(). Currently the pythonic syntactic sugar cannot generate tf.gather's but we will add that once tf.scatter is done. We will also need some broadcasting code to generate indexes as well. The advanced slicing is now a separate issue.

**Comment by xksteven** [comment]
@danijar is it documented that they have gradients? I was trying to find where its stated before I started to use them.

**Comment by danijar** [comment]
This example works for me:

```python
import numpy as np
import tensorflow as tf

variable = tf.Variable(np.ones((5, 5)))
cost = tf.reduce_sum(variable[0, :])  # Do more fancy slicing if needed.
optimize = tf.train.GradientDescentOptimizer(0.1).minimize(cost)

sess = tf.Session()
sess.run(tf.initialize_all_variables())

print('Cost (should be decreasing):')
for _ in range(5):
    print(sess.run([cost, optimize])[0])
```

**Comment by aselle** [comment]
`tf.gather_nd` does not have gradients yet, but it will soon. `tf.gather_nd` is not used by pythonic `__getitem__` slicing e.g. you cannot yet do 
`x=[1,2];y=[1,1]; foo[x,y]`.l @danijar's example is only "basic indexing". Everything that Tensor.__getiem__ supports does support Gradients.

**Comment by robsync** [comment]
Does .10 now support negative indexing? as in X[:,-1] for the last column of a tensor?  Thanks

**Comment by aselle** [comment]
0.11RC does (and also master).

**Comment by robsync** [comment]
thanks!

**Comment by guillaumeBellec** [comment]
A little inconvenience I came across.
When indexing with the Tensor I get a type error.
I will cast 'a' to int32 in the following code, but it is a bit of a hack I believe.

`import tensorflow as tf

A = tf.constant([0,2,3,1])
B = tf.constant([0,1,2,3])

a = tf.argmax(A,1)
B[a]
`

TypeError: Input 'strides' of 'StridedSlice' Op has type int32 that does not match type int64 of argument 'begin'.

**Comment by tartavull** [comment]
Is there any example on how to do gather_nd broadcasting ?

Thank you.

**Comment by ravigarg27** [comment]
Hey all, I want to update a tensor at certain locations (along one dimension only). I understand I can do it via scatter_update however it appears it doesn't have a registered gradient. Is there any workaround this?

**Comment by MInner** [comment]
@ravigarg27 a dumb workaround (if your new values are not `Variables` you're going to optimize over) is to do something conceptually like `A - bool(update!=0)*A + update`, where `update` has same shape as `A` and all not updated entries are equal to 0

**Comment by ravigarg27** [comment]
@MInner What I want is something on lines of A[indices, :] = B where A and B are Variables matrices

**Comment by MInner** [comment]
@danijar , I wonder if your [`def extract_last_relevant()`](https://github.com/tensorflow/tensorflow/issues/206#issuecomment-214560963) snipped for extracting last non-zero outputs of `tensorflow.models.rnn.dynamic_rnn`s posted above is still required (downside: `sparse to dense` warning indicating that it might lead to huge matrix allocation) or one could use new (0.11rc) partial implementation of smart indexing in tensorflow to address this issue? (I'm still little confused regarding which parts of smart indexing works now; there was an announcement in 0.11rc changelog regarding improvements in indexing, but they does not seem to address this specific "indexing via another tensor" issue, aren't they?)

**Comment by danijar** [comment]
@MInner I played around with. Unfortunately, it doesn't seem like the new indexing can simplify that.

**Comment by warmspringwinds** [comment]
Hey guys.

Sorry for bugging.
Any update on the gradient implementation for tf.gather_nd() ?

**Comment by kevinzakka** [comment]
Is there a way to do the following indexing? 

`z_i = input_img[i, x[i], y[i]]`

**Comment by danijar** [comment]
Anyone who needs to `tf.gather()` along the second dimension:

```python
def gather_along_second_dimension(data, indices):
  flat_indices = tf.tile(indices[None, :], [tf.shape(data)[0], 1])
  batch_offset = tf.range(0, tf.shape(data)[0]) * tf.shape(data)[1]
  flat_indices = tf.reshape(flat_indices + batch_offset[:, None], [-1])
  flat_data = tf.reshape(data, tf.concat([[-1], tf.shape(data)[2:]], 0))
  result_shape = tf.concat([[tf.shape(data)[0], -1], tf.shape(data)[2:]], 0)
  return tf.reshape(tf.gather(flat_data, flat_indices), result_shape)
```

**Comment by MInner** [comment]
@danijar I might have misunderstood what you are doing, but wouldn't `tf.transpose(tf.gather(tf.transpose(x, ..), ..), ..)` do a thing? Moreover, `gather_nd()` thankfully has gradients now.

**Comment by aselle** [comment]
@warmspringwinds it is deployed.

**Comment by danijar** [comment]
@MInner You're right, we can use transpose to to gather along any dimension:

```python
def gather_along_axis(data, indices, axis=0):
  if not axis:
    return tf.gather(data, indices)
  rank = data.shape.ndims
  perm = [axis] + list(range(1, axis)) + [0] + list(range(axis + 1, rank))
  return tf.transpose(tf.gather(tf.transpose(data, perm), indices), perm)
```

This is slower than my snipped above though, as it needs to rotate the whole data array in memory instead of computing the correct indices into the original tensor.

**Comment by MInner** [comment]
@danijar good to know!

**Comment by ddmbr** [comment]
Hi,

I just saw @nova77 's code above,

```python
x = tf.constant([[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]])
idx = tf.constant([1, 0, 2])
idx_flattened = tf.range(0, x.shape[0]) * x.shape[1] + idx
y = tf.gather(tf.reshape(x, [-1]),  # flatten input
              idx_flattened)  # use flattened indices

with tf.Session(''):
  print y.eval()  # [2 4 9]
```

Do we still need this workaround? Or is there already a solution (or an issue for me to track)? Thanks!

**Comment by sifatron** [comment]
Can anyone help me with this? Embedding Gather always giving Negative Index. How to solve this? Using Keras with Tensorflow backend

```
InvalidArgumentError (see above for traceback): indices[0,6] = -1 is not in [0, 15664)
	 [[Node: embedding_1/Gather = Gather[Tindices=DT_INT32, Tparams=DT_FLOAT, validate_indices=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](embedding_1/embeddings/read, embedding_1/Cast)]]
```

**Comment by epiception** [comment]
Does scatter_nd() work in the case of <unknown> indices shape? I'm trying to update a zero-filled matrix with values of dynamic indices shape. Is there a workaround for this? (Don't know if this has been asked before).

**Comment by dchatterjee172** [comment]
can anyone explain how to specifically use this axis argument?
```
>>> sess.run(glove)
array([[-0.11435109,  0.46091831,  0.32802248, -0.72811705,  0.14930534],
       [ 0.52950537, -0.72350842,  0.20540303,  0.28144443, -0.22332937],
       [ 0.03170794, -0.20141584, -0.46570423, -0.53624082, -0.38795415],
       [ 0.46187055, -0.45805836, -0.04215944, -0.65700728, -0.10993046],
       [-0.03469902,  0.36488986, -0.23033738, -0.01348984,  0.40521193]], dtype=float32)
>>> sess.run(a)
array([0, 1, 2, 3, 4], dtype=int32)
>>> bb=tf.gather(glove,a,axis=1)
>>> sess.run(bb)
array([[-0.11435109,  0.46091831,  0.32802248, -0.72811705,  0.14930534],
       [ 0.52950537, -0.72350842,  0.20540303,  0.28144443, -0.22332937],
       [ 0.03170794, -0.20141584, -0.46570423, -0.53624082, -0.38795415],
       [ 0.46187055, -0.45805836, -0.04215944, -0.65700728, -0.10993046],
       [-0.03469902,  0.36488986, -0.23033738, -0.01348984,  0.40521193]], `dtype=float32)
```
Because as you can see from the above example, it's not working.
I have to do this for the time being,
```
>>> a_=tf.stack([tf.range(0,5,1),a],axis=1)
>>> sess.run(a_)
array([[0, 0],
       [1, 1],
       [2, 2],
       [3, 3],
       [4, 4]], dtype=int32)
>>> bb=tf.gather_nd(glove,a_)
>>> sess.run(bb)
array([-0.11435109, -0.72350842, -0.46570423, -0.65700728,  0.40521193], dtype=float32)
```
